File: packages/core/src/agents/minimax-code.ts
import { createHash, type Hash } from "node:crypto";
import { homedir } from "node:os";
import { join } from "node:path";
import { getAgentCatalogEntry } from "../contract/agent-catalog.js";
import { firstExisting, readEnvPath } from "../discovery/paths.js";
import { capturePricingMisses } from "../pricing/cost.js";
import type { Message, SessionDetail, SessionHead } from "../types/index.js";
import { asString } from "../utils/narrow.js";
import { isSqliteAvailable, openDbReadOnly, type SQLiteDatabase } from "../utils/sqlite.js";
import { resolveSessionTitle } from "../utils/title-fallback.js";
import { DatabaseSessionSource, matchesScanWindow } from "./base.js";
import type { AgentScanOptions, AgentSourceOptions, SessionWatchPlan } from "./base.js";
import {
  attachMiniMaxUsage,
  miniMaxMessage,
  miniMaxUsage,
  type MiniMaxUsage,
} from "./minimax-code-content.js";

const AGENT_METADATA = getAgentCatalogEntry("minimax-code");
const DATABASE_PATH = join("v2", "sqlite", "runtime-state.sqlite");
const SESSION_FILTER =
  "(s.visibility <> 'hidden' OR s.parent_session_id IS NOT NULL) AND s.session_kind NOT IN ('peek', 'cron', 'channel')";

function dataRoots(): string[] {
  const configured = readEnvPath("MINIMAX_DATA_DIR") ?? readEnvPath("MAVIS_DATA_DIR");
  return configured
    ? [configured]
    : [join(homedir(), ".minimax"), join(homedir(), ".minimax-code")];
}

export function resolveMiniMaxCodeDataRoot(): string {
  const roots = dataRoots();
  return roots.find((root) => firstExisting(join(root, DATABASE_PATH))) ?? roots[0]!;
}

interface SessionProjection {
  head: SessionHead;
  agent: string;
  firstUserText?: string;
  signature: Hash;
  unpricedModels: Set<string>;
  messages: Message[];
  lastAssistantByTurn: Map<string, Message>;
  usageByTurn: Map<string, MiniMaxUsage[]>;
}

export class MiniMaxCodeAgent extends DatabaseSessionSource {
  readonly name = AGENT_METADATA.name;
  readonly displayName = AGENT_METADATA.displayName;
  private readonly roots: string[];

  constructor(options: AgentSourceOptions = {}) {
    super();
    this.roots = options.sourceRoot ? [options.sourceRoot] : dataRoots();
  }

  protected getDatabasePath(): string | null {
    return firstExisting(...this.roots.map((root) => join(root, DATABASE_PATH)));
  }

  isAvailable(): boolean {
    return isSqliteAvailable() && this.getDatabasePath() !== null;
  }

  getSessionWatchPlan(): SessionWatchPlan {
    return {
      status: "supported",
      targets: this.roots.flatMap((root) =>
        ["", "-wal", "-journal"].map((suffix) => ({
          path: `${join(root, DATABASE_PATH)}${suffix}`,
          pollForChanges: true,
        })),
      ),
    };
  }

  private readDatabase<T>(read: (db: SQLiteDatabase) => T): T {
    const path = this.getDatabasePath();
    if (!path) throw new Error("MiniMax Code runtime-state.sqlite is missing");
    return this.scanStep("reading MiniMax Code 0.4.12 sessions", path, () => {
      const db = openDbReadOnly(path);
      if (!db) throw new Error("Cannot open MiniMax Code database");
      try {
        return db.transaction(() => read(db))();
      } finally {
        db.close();
      }
    });
  }

  scan(options?: AgentScanOptions): SessionHead[] {
    if (!this.getDatabasePath()) return [];
    return this.readDatabase((db) => {
      const projections = this.readProjections(db);
      const selected = new Set<string>();
      const children = new Map<string, string[]>();
      for (const [id, { head }] of projections) {
        const parent = head.parent_reference?.sessionId;
        if (parent && projections.has(parent)) {
          const siblings = children.get(parent) ?? [];
          siblings.push(id);
          children.set(parent, siblings);
        } else if (matchesScanWindow(head.time_updated ?? head.time_created, options)) {
          selected.add(id);
        }
      }
      if (options?.includeRelatedSessions !== false) {
        for (const id of selected) {
          for (const child of children.get(id) ?? []) selected.add(child);
        }
      }
      const sessions: SessionHead[] = [];
      for (const id of selected) {
        const projection = projections.get(id)!;
        if (!projection.head.stats.message_count) continue;
        this.rememberSession(id, {
          sourceFingerprint: projection.signature.digest("hex"),
          headParserVersion: "minimax-code-v1",
          unpricedModels: [...projection.unpricedModels],
        });
        sessions.push(projection.head);
      }
      sessions.sort(
        (left, right) =>
          (right.time_updated ?? 0) - (left.time_updated ?? 0) ||
          left.reference.sessionId.localeCompare(right.reference.sessionId),
      );
      options?.onProgress?.({
        total: selected.size,
        processed: selected.size,
        sessions: sessions.length,
      });
      return sessions;
    });
  }

  getSessionData(sessionId: string): SessionDetail {
    return this.readDatabase((db) => {
      const projection = this.readProjections(db, sessionId).get(sessionId);
      if (!projection) throw new Error(`MiniMax Code session not found: ${sessionId}`);
      for (const [turnId, usage] of projection.usageByTurn) {
        const message = projection.lastAssistantByTurn.get(turnId);
        if (message) attachMiniMaxUsage(message, usage);
      }
      return { ...projection.head, messages: projection.messages };
    });
  }

  private readProjections(db: SQLiteDatabase, sessionId?: string): Map<string, SessionProjection> {
    const filter = `${SESSION_FILTER}${sessionId ? " AND s.session_id = ?" : ""}`;
    const bindings = sessionId ? [sessionId] : [];
    const projections = new Map<string, SessionProjection>();
    const sessions = db
      .prepare(`
      SELECT s.session_id, s.columnar_version, s.title, s.workspace_dir, s.parent_session_id,
        s.created_at_ms, s.updated_at_ms, s.agent_name, s.status, s.archived, s.extra_data_json
      FROM local_runtime_sessions s WHERE ${filter} ORDER BY s.session_id
    `)
      .iterate(...bindings);
    for (const row of sessions) {
      if (row.columnar_version !== 3)
        throw new Error("Unsupported MiniMax session columnar version");
      const id = String(row.session_id);
      const parent = asString(row.parent_session_id);
      projections.set(id, {
        head: {
          ...this.sessionIdentity(id),
          title: asString(row.title) ?? "",
          directory: asString(row.workspace_dir) ?? "",
          parent_reference:
            parent && parent !== id ? { agentName: this.name, sessionId: parent } : undefined,
          time_created: Number(row.created_at_ms ?? row.updated_at_ms),
          time_updated: Number(row.updated_at_ms),
          stats: { message_count: 0, total_input_tokens: 0, total_output_tokens: 0, total_cost: 0 },
        },
        agent: asString(row.agent_name) ?? this.displayName,
        signature: createHash("sha256").update(JSON.stringify(row)),
        unpricedModels: new Set(),
        messages: [],
        lastAssistantByTurn: new Map(),
        usageByTurn: new Map(),
      });
    }
    const messages = db
      .prepare(`
      SELECT m.id, m.session_id, m.msg_id, m.role, m.turn_id, m.source, m.created_at_ms, m.data_json
      FROM local_runtime_message_rows m JOIN local_runtime_sessions s ON s.session_id = m.session_id
      WHERE ${filter} ORDER BY m.session_id, m.id
    `)
      .iterate(...bindings);
    for (const row of messages) {
      const projection = projections.get(String(row.session_id))!;
      projection.signature.update(JSON.stringify(row));
      const message = miniMaxMessage(row, projection.agent);
      projection.head.stats.message_count += 1;
      projection.head.time_updated = Math.max(
        projection.head.time_updated ?? 0,
        message.time_created,
      );
      if (message.role === "user" && !message.automated && !projection.firstUserText) {
        projection.firstUserText = message.parts.find((part) => part.type === "text")?.text;
      }
      if (sessionId) {
        projection.messages.push(message);
        const turnId = asString(row.turn_id);
        if (turnId && message.role === "assistant" && !message.mode)
          projection.lastAssistantByTurn.set(turnId, message);
      }
    }
    const usages = db
      .prepare(`
      SELECT u.id, u.session_id, u.turn_id, u.model, u.ts, u.input_tokens, u.output_tokens,
        u.reasoning_tokens, u.cache_read_tokens, u.cache_write_tokens, u.cost_usd
      FROM local_runtime_token_usage u JOIN local_runtime_sessions s ON s.session_id = u.session_id
      WHERE ${filter} ORDER BY u.session_id, u.id
    `)
      .iterate(...bindings);
    for (const row of usages) {
      const projection = projections.get(String(row.session_id))!;
      projection.signature.update(JSON.stringify(row));
      const { result: usage, unpricedModels } = capturePricingMisses(() => miniMaxUsage(row));
      for (const model of unpricedModels) projection.unpricedModels.add(model);
      const stats = projection.head.stats;
      stats.total_input_tokens += usage.tokens.input ?? 0;
      stats.total_output_tokens += usage.tokens.output ?? 0;
      stats.total_tokens = (stats.total_tokens ?? 0) + usage.total;
      stats.total_cache_read_tokens =
        (stats.total_cache_read_tokens ?? 0) + (usage.tokens.cache_read ?? 0);
      stats.total_cache_create_tokens =
        (stats.total_cache_create_tokens ?? 0) + (usage.tokens.cache_create ?? 0);
      stats.total_cost += usage.cost;
      if (usage.costSource === "estimated" || !stats.cost_source)
        stats.cost_source = usage.costSource;
      if (usage.model) {
        const modelUsage = (projection.head.model_usage ??= Object.create(null));
        modelUsage[usage.model] = (modelUsage[usage.model] ?? 0) + usage.total;
      }
      projection.head.time_updated = Math.max(projection.head.time_updated ?? 0, usage.time);
      if (sessionId && usage.turnId) {
        const turnUsage = projection.usageByTurn.get(usage.turnId) ?? [];
        turnUsage.push(usage);
        projection.usageByTurn.set(usage.turnId, turnUsage);
      }
    }
    for (const projection of projections.values()) {
      projection.head.title = resolveSessionTitle(
        projection.head.title,
        projection.firstUserText ?? null,
        null,
      );
    }
    return projections;
  }
}

File: crates/codesesh-core/src/agents/minimax_code.rs
use super::{codex::ParsedSession, deepchat::common::*};
use crate::{contract::*, pricing::Pricing};
use anyhow::{Result, bail};
use rusqlite::{Connection, OpenFlags};
use serde_json::Value;
use std::{
    collections::{BTreeMap, BTreeSet, HashMap},
    path::Path,
};

type TurnUsage = (
    Option<String>,
    MessageTokens,
    f64,
    Option<CostSource>,
    Vec<crate::pricing::CostInput>,
);

fn convert(row: &Value, agent: &str) -> Result<Message> {
    let data = strict_parse(&row["data_json"])?;
    let id = s(&row["msg_id"])
        .filter(|id| !id.is_empty())
        .ok_or_else(|| anyhow::anyhow!("Invalid MiniMax display message"))?;
    if !data.is_object() || data["msg_id"].as_str() != Some(&id) {
        bail!("Invalid MiniMax display message");
    }
    let role = match row["role"]
        .as_str()
        .or(data["role"].as_str())
        .unwrap_or("assistant")
    {
        "user" => Role::User,
        "assistant" => Role::Assistant,
        other => bail!("Invalid MiniMax role: {other}"),
    };
    let time = row["created_at_ms"]
        .as_f64()
        .or(data["timestamp"].as_f64())
        .unwrap_or(0.0);
    let mut parts = Vec::new();
    let kind = s(&data["kind"]);
    if let Some(kind) = kind.as_ref().filter(|kind| !kind.is_empty()) {
        let label = match kind.as_str() {
            "compaction_start" => "Context compaction started",
            "compaction" => "Context compacted",
            "compaction_failed" => "Context compaction failed",
            "review_start" => "Review started",
            "review_result" => "Review result",
            "review_failed" => "Review failed",
            "review_aborted" => "Review aborted",
            "review_interrupted" => "Review interrupted",
            _ => "",
        };
        parts.push(text(
            if label.is_empty() {
                format!("Event: {kind}")
            } else {
                label.into()
            },
            None,
        ));
    }
    if let Some(value) = s(&data["thinking_content"]).filter(|s| !s.is_empty()) {
        parts.push(MessagePart::Reasoning {
            text: value,
            time_created: None,
        });
    }
    if let Some(value) = s(&data["msg_content"]).filter(|s| !s.is_empty()) {
        parts.push(text(value, None));
    }
    for attachment in data["attachments"].as_array().into_iter().flatten() {
        let name = ["file_path", "desktop_path", "file_name", "asset_id", "url"]
            .into_iter()
            .find_map(|key| s(&attachment[key]));
        if let Some(name) = name.filter(|s| !s.is_empty()) {
            let mime = s(&attachment["mime_type"])
                .filter(|s| !s.is_empty())
                .map(|mime| format!(" ({mime})"))
                .unwrap_or_default();
            parts.push(text(format!("Attachment: {name}{mime}"), None));
        }
    }
    let mut children = BTreeSet::new();
    for call in data["tool_calls"].as_array().into_iter().flatten() {
        let Some(tool) = s(&call["tool_name"]).filter(|s| !s.is_empty()) else {
            continue;
        };
        let output = parse(&call["tool_call_result_data"]);
        let failed = call["tool_call_status"] == 3
            || output["isError"] == true
            || !output["error"].is_null();
        let mut metadata = output["details"].as_object().cloned().unwrap_or_default();
        if let Some(status) = call.get("tool_call_status") {
            metadata.insert("minimax_status".into(), status.clone());
        }
        if (tool == "task" || tool.starts_with("task_"))
            && let Some(child) = metadata
                .get("sub_session_id")
                .and_then(Value::as_str)
                .or_else(|| metadata.get("session_id").and_then(Value::as_str))
        {
            children.insert(child.to_owned());
        }
        parts.push(MessagePart::Tool {
            tool,
            call_id: s(&call["tool_call_id"]),
            title: None,
            time_created: None,
            state: Box::new(ToolState {
                status: if failed {
                    "error"
                } else if call["tool_call_status"] == 2 {
                    "completed"
                } else {
                    "running"
                }
                .into(),
                input: call.get("tool_call_args").map(parse),
                output: call.get("tool_call_result_data").map(parse),
                error: failed.then(|| {
                    output
                        .get("error")
                        .filter(|v| !v.is_null())
                        .or_else(|| output.get("text").filter(|v| !v.is_null()))
                        .unwrap_or(&output)
                        .clone()
                }),
                metadata: Some(Value::Object(metadata)),
            }),
        });
    }
    if let Some(error) = s(&data["error"]).filter(|s| !s.is_empty()) {
        parts.push(text(error, None));
    }
    let automated = role == Role::User
        && (kind.as_ref().is_some_and(|s| !s.is_empty())
            || ["cron", "thread-goal", "team", "background-task", "system"]
                .contains(&row["source"].as_str().unwrap_or("")));
    let mut message = message(id, role, time, parts);
    if message.role == Role::Assistant {
        message.agent = Some(agent.into());
    }
    if data
        .get("finish_reason")
        .is_some_and(|v| !v.is_null() && v != false && v != "" && v != 0)
    {
        message.time_completed = Some(time);
    }
    message.mode = kind;
    message.automated = automated.then_some(true);
    message.subagent_id = (children.len() == 1).then(|| children.into_iter().next().unwrap());
    Ok(message)
}

pub fn scan(root: &Path, pricing: &Pricing) -> Result<Vec<ParsedSession>> {
    scan_impl(root, pricing, None)
}
pub fn scan_selected(
    root: &Path,
    pricing: &Pricing,
    selected: &std::collections::HashSet<String>,
) -> Result<Vec<ParsedSession>> {
    scan_impl(root, pricing, Some(selected))
}
fn scan_impl(
    root: &Path,
    pricing: &Pricing,
    selected: Option<&std::collections::HashSet<String>>,
) -> Result<Vec<ParsedSession>> {
    let path = root.join("v2/sqlite/runtime-state.sqlite");
    if !path.exists() {
        return Ok(Vec::new());
    }
    let db = Connection::open_with_flags(&path, OpenFlags::SQLITE_OPEN_READ_ONLY)?;
    db.execute_batch("BEGIN")?;
    scan_connection(root, &path, &db, pricing, selected)
}
fn scan_connection(
    _root: &Path,
    path: &Path,
    db: &Connection,
    pricing: &Pricing,
    selected: Option<&std::collections::HashSet<String>>,
) -> Result<Vec<ParsedSession>> {
    let read = |sql: &str| rows(db, &selected_query(sql, "s.session_id", selected));
    let filter = "(s.visibility <> 'hidden' OR s.parent_session_id IS NOT NULL) AND s.session_kind NOT IN ('peek', 'cron', 'channel')";
    let headers = rows(
        db,
        &format!("SELECT s.* FROM local_runtime_sessions s WHERE {filter} ORDER BY s.session_id"),
    )?;
    let mut all_messages: HashMap<String, Vec<Value>> = HashMap::new();
    for row in read(&format!(
        "SELECT m.* FROM local_runtime_message_rows m JOIN local_runtime_sessions s ON s.session_id=m.session_id WHERE {filter} ORDER BY m.session_id,m.id"
    ))? {
        all_messages
            .entry(s(&row["session_id"]).unwrap_or_default())
            .or_default()
            .push(row);
    }
    let mut all_usage: HashMap<String, Vec<Value>> = HashMap::new();
    for row in read(&format!(
        "SELECT u.* FROM local_runtime_token_usage u JOIN local_runtime_sessions s ON s.session_id=u.session_id WHERE {filter} ORDER BY u.session_id,u.id"
    ))? {
        all_usage
            .entry(s(&row["session_id"]).unwrap_or_default())
            .or_default()
            .push(row);
    }
    let mut result = Vec::new();
    let reachable = reachable_sessions(headers.iter().map(|h| {
        let id = h["session_id"].as_str().unwrap_or("");
        (
            id,
            h["parent_session_id"]
                .as_str()
                .filter(|p| !p.is_empty() && *p != id),
        )
    }));
    for header in headers {
        if selected.is_some_and(|ids| !ids.contains(header["session_id"].as_str().unwrap_or(""))) {
            continue;
        }
        if header["columnar_version"] != 3 {
            bail!("Unsupported MiniMax session columnar version");
        }
        let id = s(&header["session_id"]).unwrap_or_default();
        let agent = s(&header["agent_name"]).unwrap_or_else(|| "MiniMax Code".into());
        let mut updated = n(&header["updated_at_ms"]);
        let mut messages = Vec::new();
        let mut last_by_turn = HashMap::new();
        let mut first_user = None;
        for row in all_messages.remove(&id).unwrap_or_default() {
            let message = convert(&row, &agent)?;
            updated = updated.max(message.time_created);
            if message.role == Role::User && message.automated != Some(true) && first_user.is_none()
            {
                first_user = message.parts.iter().find_map(|p| {
                    if let MessagePart::Text { text, .. } = p {
                        Some(text.clone())
                    } else {
                        None
                    }
                });
            }
            if message.role == Role::Assistant
                && message.mode.as_ref().is_none_or(|mode| mode.is_empty())
                && let Some(turn) = s(&row["turn_id"])
            {
                last_by_turn.insert(turn, messages.len());
            }
            messages.push(message);
        }
        let mut stats = SessionStats {
            message_count: messages.len(),
            ..Default::default()
        };
        let mut model_usage = BTreeMap::new();
        let mut usage_by_turn: HashMap<String, Vec<TurnUsage>> = HashMap::new();
        for row in all_usage.remove(&id).unwrap_or_default() {
            let model = s(&row["model"]).filter(|s| !s.is_empty());
            let tokens = MessageTokens {
                input: Some(nonnegative(&row["input_tokens"])),
                output: Some(nonnegative(&row["output_tokens"])),
                reasoning: Some(nonnegative(&row["reasoning_tokens"])),
                cache_read: Some(nonnegative(&row["cache_read_tokens"])),
                cache_create: Some(nonnegative(&row["cache_write_tokens"])),
            };
            let total = [
                tokens.input,
                tokens.output,
                tokens.reasoning,
                tokens.cache_read,
                tokens.cache_create,
            ]
            .into_iter()
            .flatten()
            .sum();
            let recorded = row["cost_usd"].as_f64().filter(|n| *n >= 0.0);
            let mut cost_inputs = Vec::new();
            let cost = recorded.or_else(|| {
                pricing.estimate_tracked(model.as_deref(), &tokens, 0.0, &mut cost_inputs)
            });
            stats.cost_inputs.extend(cost_inputs.iter().cloned());
            let source = if recorded.is_some() {
                Some(CostSource::Recorded)
            } else {
                cost.map(|_| CostSource::Estimated)
            };
            add_stats(
                &mut stats,
                &tokens,
                total,
                cost.unwrap_or(0.0),
                source.clone(),
            );
            if let Some(model) = &model {
                *model_usage.entry(model.clone()).or_insert(0.0) += total;
            }
            updated = updated.max(nonnegative(&row["ts"]));
            if let Some(turn) = s(&row["turn_id"]) {
                usage_by_turn.entry(turn).or_default().push((
                    model,
                    tokens,
                    cost.unwrap_or(0.0),
                    source,
                    cost_inputs,
                ));
            }
        }
        for (turn, usages) in usage_by_turn {
            let Some(index) = last_by_turn.get(&turn) else {
                continue;
            };
            let models: BTreeSet<_> = usages.iter().map(|u| u.0.clone()).collect();
            if models.len() != 1 {
                continue;
            }
            let message = &mut messages[*index];
            message.cost = Some(0.0);
            message.model = usages[0].0.clone();
            let mut tokens = empty_tokens();
            for (_, usage, cost, source, cost_inputs) in usages {
                message.cost_inputs.extend(cost_inputs);
                add_tokens(&mut tokens, &usage);
                message.cost = Some(message.cost.unwrap_or(0.0) + cost);
                if source == Some(CostSource::Estimated) || message.cost_source.is_none() {
                    message.cost_source = source;
                }
            }
            message.tokens = Some(tokens);
        }
        if messages.is_empty() || !reachable.contains(&id) {
            continue;
        }
        let parent =
            s(&header["parent_session_id"]).filter(|parent| !parent.is_empty() && parent != &id);
        result.push(finish(
            path,
            "minimax-code",
            id,
            title(&[header["title"].as_str(), first_user.as_deref()]),
            s(&header["workspace_dir"]).unwrap_or_default(),
            header["created_at_ms"]
                .as_f64()
                .unwrap_or(n(&header["updated_at_ms"])),
            updated,
            parent,
            stats,
            (!model_usage.is_empty()).then_some(model_usage),
            messages,
        ));
    }
    result.sort_by(|a, b| {
        b.detail
            .head
            .time_updated
            .total_cmp(&a.detail.head.time_updated)
            .then(crate::locale::compare(
                &a.detail.head.reference.session_id,
                &b.detail.head.reference.session_id,
            ))
    });
    Ok(result)
}

pub fn fingerprints(root: &Path) -> Result<HashMap<String, String>> {
    let path = root.join("v2/sqlite/runtime-state.sqlite");
    if !path.exists() {
        return Ok(HashMap::new());
    }
    let db = Connection::open_with_flags(path, OpenFlags::SQLITE_OPEN_READ_ONLY)?;
    db.execute_batch("BEGIN")?;
    fingerprint_connection(&db, None)
}
fn fingerprint_connection(
    db: &Connection,
    selected: Option<&std::collections::HashSet<String>>,
) -> Result<HashMap<String, String>> {
    let filter = "(s.visibility <> 'hidden' OR s.parent_session_id IS NOT NULL) AND s.session_kind NOT IN ('peek', 'cron', 'channel')";
    super::deepchat::common::selected_fingerprints(
        db,
        &[
            (
                &format!(
                    "SELECT s.* FROM local_runtime_sessions s WHERE {filter} ORDER BY s.session_id"
                ),
                "session_id",
            ),
            (
                &format!(
                    "SELECT m.* FROM local_runtime_message_rows m JOIN local_runtime_sessions s ON s.session_id=m.session_id WHERE {filter} ORDER BY m.session_id,m.id"
                ),
                "session_id",
            ),
            (
                &format!(
                    "SELECT u.* FROM local_runtime_token_usage u JOIN local_runtime_sessions s ON s.session_id=u.session_id WHERE {filter} ORDER BY u.session_id,u.id"
                ),
                "session_id",
            ),
        ],
        "s.session_id",
        selected,
    )
}

pub fn scan_page(
    root: &Path,
    pricing: &Pricing,
    selected: &std::collections::HashSet<String>,
) -> Result<(Vec<ParsedSession>, HashMap<String, String>)> {
    let path = root.join("v2/sqlite/runtime-state.sqlite");
    if !path.exists() {
        return Ok((Vec::new(), HashMap::new()));
    }
    let db = Connection::open_with_flags(&path, OpenFlags::SQLITE_OPEN_READ_ONLY)?;
    db.execute_batch("BEGIN")?;
    let fingerprints = fingerprint_connection(&db, Some(selected))?;
    let sessions = scan_connection(root, &path, &db, pricing, Some(selected))?;
    Ok((sessions, fingerprints))
}

pub fn enumerate_session_keys(root: &Path) -> Result<Vec<(String, f64)>> {
    let path = root.join("v2/sqlite/runtime-state.sqlite");
    if !path.exists() {
        return Ok(Vec::new());
    }
    let db = Connection::open_with_flags(path, OpenFlags::SQLITE_OPEN_READ_ONLY)?;
    db.execute_batch("BEGIN")?;
    let headers = rows(
        &db,
        "SELECT s.session_id,s.parent_session_id,s.columnar_version,COALESCE(m.message_count,0) AS message_count,MAX(COALESCE(s.updated_at_ms,0),COALESCE(m.activity,0),COALESCE(u.activity,0)) AS activity FROM local_runtime_sessions s LEFT JOIN (SELECT session_id,COUNT(*) AS message_count,MAX(created_at_ms) AS activity FROM local_runtime_message_rows GROUP BY session_id) m ON m.session_id=s.session_id LEFT JOIN (SELECT session_id,MAX(ts) AS activity FROM local_runtime_token_usage GROUP BY session_id) u ON u.session_id=s.session_id WHERE (s.visibility <> 'hidden' OR s.parent_session_id IS NOT NULL) AND s.session_kind NOT IN ('peek', 'cron', 'channel')",
    )?;
    if headers.iter().any(|h| h["columnar_version"] != 3) {
        bail!("Unsupported MiniMax session columnar version");
    }
    let selected = reachable_sessions(headers.iter().map(|h| {
        let id = h["session_id"].as_str().unwrap_or("");
        (
            id,
            h["parent_session_id"]
                .as_str()
                .filter(|p| !p.is_empty() && *p != id),
        )
    }));
    let mut keys: Vec<_> = headers
        .into_iter()
        .filter_map(|h| {
            let id = s(&h["session_id"])?;
            (n(&h["message_count"]) > 0.0 && selected.contains(&id))
                .then(|| (id, n(&h["activity"])))
        })
        .collect();
    keys.sort_by(|a, b| {
        b.1.total_cmp(&a.1)
            .then_with(|| crate::locale::compare(&a.0, &b.0))
    });
    Ok(keys)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    #[test]
    fn display_events_and_tool_metadata() {
        let row = json!({"msg_id":"m","role":"user","source":"cron","created_at_ms":42,"data_json":json!({"msg_id":"m","msg_content":"hello","tool_calls":[{"tool_name":"task","tool_call_status":3,"tool_call_result_data":"{\"error\":\"failed\",\"details\":{\"sub_session_id\":\"child\"}}"}]}).to_string()});
        let m = convert(&row, "MiniMax Code").unwrap();
        assert_eq!(m.automated, Some(true));
        assert_eq!(m.subagent_id.as_deref(), Some("child"));
        assert!(matches!(&m.parts[1],MessagePart::Tool{state,..} if state.status=="error"));
    }
    #[test]
    fn mismatched_message_identity_is_rejected() {
        assert!(
            convert(
                &json!({"msg_id":"a","data_json":"{\"msg_id\":\"b\"}"}),
                "agent"
            )
            .is_err()
        );
    }
}
#[cfg(test)]
#[path = "minimax_code/scan_tests.rs"]
mod scan_tests;
